### DISCUSSION TITLE
OneDrive Client Version v2.5.0-alpha-1 is now available for testing

### DISCUSSION THREAD

**Discussion by abraunegg** [abraunegg/onedrive#2495] [General]
As part of the ongoing development of v2.5.0, client version v2.5.0-alpha-1 is now available for testing.

**Changelog from alpha-0:**
* Added support for OneDrive Business Shared Folders.
  * This is a 100% re-write of this capability which changes how this was originally implemented to support [FR-1251](https://github.com/abraunegg/onedrive/issues/1251) , [BUG-2379](https://github.com/abraunegg/onedrive/issues/2379) and [BUG-2434](https://github.com/abraunegg/onedrive/issues/2434)
  * Please read the updated documentation: https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-alpha-1/docs/business-shared-folders.md

Your requirements will be, should you wish to help test:
* You MUST use a OneDrive TEST Account
  * OneDrive Personal Accounts are supported - create a new account here: https://www.microsoft.com/en-au/microsoft-365/onedrive/online-cloud-storage
  * OneDrive Personal Shared Folders are supported
  * OneDrive Business Accounts are supported
  * OneDrive Business Shared Folders are supported. Please read the updated documentation.
* Do not use your normal OneDrive configuration directory (your normal 'sync_dir') - set a 100% different 'sync_dir'
 * You MUST be using test data. Please do not use any alpha-application-code with your *actual* data or data you care about.
* SharePoint Libraries are hard coded at the moment to be not supported.

You can use your *normal* account if you wish (this is your choice), however this is 100% alpha quality software at the moment. If you experience data loss or something else with your real data - whilst not good, please take every effort to *not* use this code with your *real* data.

My ask is that you create a test report - what works, what does not work, what could be potentially changed | improved - basically I am soliciting some feedback on where this total re-write is heading as it matures to a v2.5.0 release.

### How can you get v2.5.0-alpha-1 ?

You need to build this client from source. Follow the high level directions here to get the development requirements installed first: https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#building-from-source---high-level-requirements

Next clone and checkout the PR:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/2494/head:pr2494
git checkout pr2494
./configure; make clean; make;
```

To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

When running the PR, your version should be as per below or greater:
```
v2.5.0-alpha-1 GitHub version: v2.4.25-48-g4bd9ae5
```
The important point is `v2.5.0-alpha-1`. The `v2.4.25-XX-abcdefg` will change as new commits are put into the PR branch, whilst things are added/fixed/updated.

**What is known to be not functioning right at this stage:**
* Application logging output when downloading & uploading files is jumbled. The reason for this is all the parallel threads writing to terminal at the same time. This is even more evident when downloading or uploading files >4MB. This will not be fixed in `alpha-1` but will need to be focused upon for future releases once feature parity with v2.4.25 is reached. Please do not raise an issue ticket because of this.
* Existing documentation *does not match* what v2.5.0 does or supports at present. It will be updated as development progresses. 

### I have no test files or test data - what can I use?
In developing this application I use a script as per below to generate random data:
```
#!/bin/bash
FILECOUNT=0
DIRCOUNT=0
DESTDIR="/my/path/to/my/sync_dir/folder/OneDrive/random_100_files"
FILES=10
FOLDERS=10
BLOCKSIZE=4

mkdir -p $DESTDIR

while [ $DIRCOUNT -lt $FOLDERS ] ; do
        DIRNAME=`cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1`
        DIRPATH=$DESTDIR/$DIRNAME
        mkdir -p $DIRPATH
        while [ $FILECOUNT -lt $FILES ] ; do
                FILESIZE=$RANDOM
                FILESIZE=$(($FILESIZE+1024))
                FILENAME=$DIRPATH/file${FILECOUNT}.data
                echo "Creating $FILENAME"
                dd if=/dev/urandom of=${FILENAME} bs=$BLOCKSIZE count=$FILESIZE > /dev/null 2>&1
                ((FILECOUNT++))
        done
        FILECOUNT=0
        ((DIRCOUNT++))
done
```

Update `DESTDIR` with the right directory and adjust `FILES` and `FOLDERS` as desired. The above settings will create 10 folders, with each folder having 10 files. with each file size between 6KB and 125KB.

To increase (or decrease) the random file size, change `BLOCKSIZE`, generally use ^2 values: 1,2,4,8,16,32,64,128,256 and so on.

At 128 you will get files from 100KB -> ~3.8MB, at 256 most files will be between 500KB --> 7.5MB

**Comment by zaxlofful** [comment]
Is there an ETA, for when this will hit the live version?

**Reply by abraunegg** [reply]
@ZaxLofful 
Please catchup with where things are at with alpha-2
